### PR TITLE
refactor(store): modélise/teste la création d'un bouquet

### LINKED ISSUE #92: UI rendering in composables
### What I did

I tried the following code:

```js
import { beforeEach, expect, test } from 'vitest'
import { setActivePinia, createPinia } from 'pinia'
import { useBouquetDatasetStore } from '../BouquetDatasetStore'

beforeEach(() => {
  setActivePinia(createPinia())
})

test('fetch a dataset', async () => {
  const id = 'XXX'
  const store = useBouquetDatasetStore()
  const dataset = await store.get(id)
  expect(dataset.name).toMatch("Données de consommation et de points")
  expect(dataset.uri).toMatch("https://demo.data.gouv.fr/fr/datasets/")
})
```

### What I expected to happen

To work:

```
✓ src/store/__test__/BouquetDatasetStore.test.js (1) 472ms
   ✓ fetch a dataset 472ms

 Test Files  1 passed (1)
      Tests  1 passed (1)
```

### What actually happened

It failed with the following error:

```
ReferenceError: document is not defined
 ❯ Object.show node_modules/vue-loading-overlay/dist/index.js:517:9
    515|       let container = propsData.container;
    516|       if (!propsData.container) {
    517|         container = document.body;
       |         ^
    518|         propsData.isFullPage = true;
    519|       }
 ❯ DatasetsAPI.makeRequestAndHandleResponse src/services/api/DatagouvfrAPI.js:62:29
 ❯ DatasetsAPI.get src/services/api/DatagouvfrAPI.js:80:23
 ❯ Proxy.get src/store/BouquetDatasetStore.js:20:47
 ❯ Proxy.<anonymous> node_modules/pinia/dist/pinia.mjs:1367:30
 ❯ src/store/__test__/BouquetDatasetStore.test.js:10:31
```

### Data (or links to it) that can help you reproduce this issue

It seems that the [DatagouvfrAPI](https://github.com/ecolabdata/ecospheres-front/blob/main/src/services/api/DatagouvfrAPI.js#L61) composable is directly rendering UI elements:

```js
import { useLoading } from "vue-loading-overlay"
...

  async makeRequestAndHandleResponse (url, method = "get", params = {}) {
    const loader = $loading.show()  // <-- here

    return this.request(url, method, params).catch(error => {
      console.error(error)
      if (error && error.message) {
        toast(error.message, { type: "error", autoClose: false }) // <-- here
      }
    }).finally(() => loader.hide()) // <-- here
  }
```

However, it seems more desirable to be up to the components to decide what and how to render based on the information provided by composables, to have UI isolation and a better separation between UI and the rest.

Otherwise, to make changes in the front about what message provide to the user in case of an error, one would need to make changes to the composable logic (handling communication with the API).

For example, in the case of the loader depicted above, the followings seems preferable:

```js
import { ref } from 'vue'

  const loading = ref(false)
  ...

  async makeRequestAndHandleResponse (url, method = "get", params = {}) {
    loading.value = true  // <-- UI logic factored out
    ...
  }
```

### LINKED ISSUE #99: refactor(store): modéliser/tester la création d'un bouquet
Superceded by #144
Superceded by #162 
Superceded by #178 

### Job story

> User: Tous.
>
> When we're about to launch "Écosphères Bêta",
> we want to create data bouquets,
> so to encourage data re-use and discoverability by use case.

### Context or situation

Quand j'implemente la création des bouquets, je m'attends à ce qu'il y ait une séparation claire entre les composants et l'API DataGouvFR et qu'il y ait des tests d'intégration qui permettent de comprendre la logique derrière.

### Problem encountered by users

Aucun des deux n'existe pour le moment.

### Proposal of how to solve the problem

- Ajouter un/des stores pour la création d'un bouquet.
- Ajouter de tests d'intégration.

### Additional elements of my proposal

N/A.

### Definition of done

- La sépation est claire en lissant le code.
- Les changements sont invisibles pour l'utilisateur
- Les stores modélisent à minimna un `Bouquet`, un `Information`, un `Datause`, et un `Dataset`.
- Les tests existent et passent en local.

### Related OKR

N/A.

### PR DESCRIPTION
**Pull request by bonjourmauko**
Closes #92
Closes #99 
Relates to #103
Depends on #91
Depends on #90 

### Job story

> User: Tous.
>
> When we're about to launch "Écosphères Bêta",
> we want to create data bouquets,
> so to encourage data re-use and discoverability by use case.

### Context or situation

Quand j'implemente la création des bouquets, je m'attends à ce qu'il y ait une séparation claire entre les composants et l'API DataGouvFR et qu'il y ait des tests unitaires et/ou d'intégration qui permettent de comprendre la logique derrière.

### Problem encountered by users

Aucun des deux n'existe pour le moment.

### Proposal of how to solve the problem

- Ajoute un stores pour la création d'un bouquet.
- Ajoute des tests unitaires et d'intégration.

### Additional elements of my proposal

N/A.

### CODE REVIEW

File: src/store/__tests__/createBouquet.test.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+beforeEach(async (context) => {
+  setActivePinia(createPinia())
+  const client = new TopicsAPI()
+  const storeFactory = createBouquetStore(client)
+  context.client = client
+  context.store = storeFactory()
+})
+
+afterEach(async ({ client, store }) => {
```
Comment (maukoquiroga) [inline src/store/__tests__/createBouquet.test.js:16]: poke @abulte @streino

File: src/contexts/createBouquet/datasetProperties.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
+/**
```
Comment (maukoquiroga) [inline src/contexts/createBouquet/datasetProperties.js:1]: Par exemple ici, on pourrait rajouter une documentation avec des examples :

```js
\**
* Dataset properties give additional context to each dataset in a bouquet.
* 
* Imagine the following `dataset`:
* 
* ```json
* {
*   "id": "1a2s3d4f",
*   "title": "Données thermofluviales",
*   "description": "Description qui ne donne pas indice du contexte d'utilisation de la donnée"
* }
* ```
*
* If we add that dataset to a `bouquet` (`topic`), we'd still lack that context, to know what the data can be 
* used for:
*
* ```json
* {
*   "title": "Indicateurs thermofluviaux",
*   "description": "...",
*   "tags": ["ecolab", "theme:agir"],
*   "datasets" : [
*     {
*       "id": "1a2s3d4f",
*       "title": "Données thermofluviales",
*       "description": "Description qui ne donne pas indice du contexte d'utilisation de la donnée"
*     }
*   ]
* }
* ```
*
* In order to give more context, we use an intermediate entity called {DatasetProperties}. For example, to complete 
* our previous example, imagine that the dataset comes from `data.gouv.fr`:
* 
* ```json
* {
*   "title": "Indicateurs thermofluviaux",
*   "description": "...",
*   "tags": ["ecolab", "theme:agir"],
*   "datasets" : [
*     {
*       "id": "1a2s3d4f",
*       "title": "Données thermofluviales",
*       "description": "Description qui ne donne pas indice du contexte d'utilisation de la donnée"
*     }
*   ],
*   "datasets_properies": [
*     {
*       "name": "The name of the properties (not the dataset)",
*       "description": "Why is this dataset associated with the bouquet",
*       "datagouvId": "1a2s3d4f"
*     }
*   ]
* }
* 
* Etc.
* 
* ```
*/
```

Comment (maukoquiroga) [reply src/contexts/createBouquet/datasetProperties.js:1]: poke @YeLnatSs @streino et al.

**Review by abulte — COMMENTED** [review]
Yay for testing (in any case) and domain modeling (in principle).

Still, I feel like this is overly complicated given the current scope of the project. I understand that this is meant to be future and scale proof but maybe we could do this in a leaner/incremental way.

I only glanced at the PR, given its huge size, and I hope I understood what you're trying to do. Some general comments below.

## Data model

It feels like you're trying to build a "model management" framework, quite similar to [marshmallow in python](https://github.com/marshmallow-code/marshmallow). 

I looked for such libraries in JS (I don't know any from the top of my head), and it's interesting that I only found abandoned projects: https://www.npmjs.com/package/datamodeljs, https://vuemc.io/, https://github.com/benpickles/js-model...

I think this is because the crux of the effort in JS has moved to state management, and marshmallow is solving a problem that Javascript does not have : convert from/to Python to/from JSON. JSON is ofc native to JS. Still, the fact that I did not find (maybe I did not look well enough) a library to take off some the boilerplate you added is for me a signal that we're maybe trying to solve a problem that "normal people" don't have.

On a side note, [udata admin has a concept of model](https://github.com/opendatateam/udata/blob/master/js/models/base.js) and I've worked with it for many years. On paper it's great, the models are even defined from the Swagger documentation! The model handles the API calls transparently! In practice, it's been a nightmare to work with, even though the scope of data.gouv.fr admin could warrant such a pattern (which Ecospheres does not IMHO). Too many abstractions to reason about, very difficult to debug and maintain.

## createBouquet

I do not really understand why you're working on a "createBouquet" scope. I feel like the scope should be the Bouquet, with sub actions / logic related to creation. The fact that you're creating a store dedicated to `createBouquet` feels wrong in a piniaesque Vue-world, the store should be the BouquetStore, maybe with sub-stores if stuff get really messy. I also feel like naming / scoping things with `create` make it hard to reason about, because you don't know the impacts on the rest of CRUD operations.

## Proposal

Maybe add the thinnest layer possible on top of a simple object (dataset extras?), to demonstrate what you're trying to achieve with basic building blocks in a simple PR. This would make the added value simpler to grasp, help us find patterns / libraries if we need some, and maybe cut away some of the cruft when it's generalized.

Maybe it' also a bad idea to mix introducing testing along with data models. I understand why they make sense together, but tackling them at the same time feels overwhelming. Maybe separate the efforts at first.

**Comment by YeLnatSs** [comment]
this kind of example is available ?

 {
   "title": "Name of Bouquet",
   "description":  Description of the Bouquet",
  "theme": "Produire",
  "subtheme": "Nucléaire",
   "tags":  ["ecolab", "theme:agir"],
   "datasets" :  [
    {
       "id": "1a2s3d4f",
       "title": "Populations et territoires exposés au dépassement de la valeur limite des PM10 sur la région Grand-Est pour l'année 2018",
       "description": "Populations et territoires exposés au dépassement de la valeur limite des PM10 (35 jours supérieurs à 50 µg/m3) sur la région Grand-Est pour l'année 2018. L'indicateur de dépassement de la valeur limite de 50 µg/m3 ...",
         "uri": "https://demo.data.gouv.fr/api/1/datasets/base-des-permis-de-construire-et-autres-autorisations-durbanisme-sitadel/"
     }
   ],
   "datasets_properies":  [
     {
      "name": "The text inside the libelle input",
      "description": "The text we had inside the raison textarea",
       "datagouvId": "6537913f29a7934b25a884db",
       "uri": "https://demo.data.gouv.fr/api/1/datasets/base-des-permis-de-construire-et-autres-autorisations-durbanisme-sitadel/"
     }
   ]
}

**Comment by maukoquiroga** [comment]
@abulte Just wanted to fix #92, the rest is purely a TDD accident: let the store just carry-over state mutations, not much more. More generally, I don't get mixing presentation and data. Looks cool by feels sloppy (cf. #92). Thanks indeed!

**Comment by maukoquiroga** [comment]
(BTW, the "model management" thing was not on purpose: see the models were extracted from the stores in the last 3 or 4 commits, before all the representation logic was there as well).

File: src/contexts/createBouquet/datasetProperties.js (hunk 2 above)
Comment (streino) [reply src/contexts/createBouquet/datasetProperties.js:1]: J'ai l'impression que ces exemples peuvent être des tests (et donc du code) plutôt que de la doc, non ? L'avantage c'est que les tests risquent moins de devenir incohérents avec le code que la doc.

**Comment by streino** [comment]
Globalement +1 sur les remarques de @abulte.

Je ne connais pas l'écosystème JS mais j'ai tout de même l'impression qu'on devrait pouvoir atteindre nos objectifs plus simplement. Objectifs qui sont, si j'ai bien compris (en parenthèses mon évaluation) : tests (requis) + modélisation (souhaitable, en partie nécessaire pour les tests) + découplage (découle de l'amélioration de la modélisation).

Je ne comprends en particulier pas pourquoi on élabore ici un modèle en plus des stores existants.
Et je m'interroge sur la complexité de la Serde JS-JSON.

Il me semble que procéder par petites PR incrémentales qu'on peut merger rapidement simplifierait l'interaction autour de ce(s) gros sujet(s). A voir s'il est plus simple de remodeler un peu l'existant avant de pouvoir ajouter qqs tests minimaux, ou l'inverse ?

**Comment by maukoquiroga** [comment]
> Je ne connais pas l'écosystème JS mais j'ai tout de même l'impression qu'on devrait pouvoir atteindre nos objectifs plus simplement. Objectifs qui sont, si j'ai bien compris (en parenthèses mon évaluation) : tests (requis) + modélisation (souhaitable, en partie nécessaire pour les tests) + découplage (découle de l'amélioration de la modélisation).

Tests d'abord AMHA. En ce cette PR fait un poil trop.

> Je ne comprends en particulier pas pourquoi on élabore ici un modèle en plus des stores existants.
Et je m'interroge sur la complexité de la Serde JS-JSON.

Juste pour avoir la discussion :) sur « ce quoi le domaine ». Qu'il ait un « TopicStore » maintenant au lieu d'un « BouquetStore » parle des choix de domaine implicites et (encore) non documentés.

(Il ne faut pas confondre « créer un modèle logiciel » avec modéliser le domaine métier). 

Parce que le domaine metier c'est les bouquets, même si derrière dans data.gouv.fr c'est un mélange de topics et datasets, ici cela devrait s'apperler bouquets. Ici le domaine, les modèles donc, devrient être un mirroir de la UI.

> Et je m'interroge sur la complexité de la Serde JS-JSON.

Cette PR fait beaucoup d'opérations inutiles en réalité. Il s'agit simplement de JS/JSON-to-JS/JSON, de transformer ce qui nous dicte la UI en quelque chose de comprehensible pour l'API.

> A voir s'il est plus simple de remodeler un peu l'existant avant de pouvoir ajouter qqs tests minimaux, ou l'inverse ?

C'est un peu l'oeuf et la poule 😄

**Comment by maukoquiroga** [comment]
One picture is worth a thousand LOCs :

**Comment by abulte** [comment]
For reference, if we ever need them (not sure at this point), I found two interesting libraries for schema/model definition/validation:
- https://www.npmjs.com/package/yup
- https://www.npmjs.com/package/joi

**Comment by maukoquiroga** [comment]
Thanks @abulte :)

Connais-tu https://github.com/typestack/class-transformer ?

**Comment by bonjourmauko** [comment]
Je ferme car les idées contenues ici sont déjà en grande partie reprises : 
- les tests #138 #52 #32 
- la modélisation #81 #32 #145

**Pull request closed without merging**